packages/compass-connections-navigation/src/connections-navigation-tree.tsx

```diff
@@ -0,0 +1,541 @@
+            )
+          );
+        }}
+        fallback={() => (
+          <PlaceholderItem
+            level={itemData.level}
+            isLegacy={isLegacy}
+          ></PlaceholderItem>
+        )}
+      ></FadeInPlaceholder>
+    </div>
+  );
+}, areEqual);
+
+const navigationTree = css({
+  flex: '1 0 auto',
+});
+
+const ConnectionsNavigationTree: React.FunctionComponent<{
+  connections?: Connection[];
+  databasesLegacy?: Database[];
+  expanded?: Record<string, false | Record<string, boolean>>;
+  expandedLegacy?: Record<string, boolean>;
+  onConnectionExpand?(id: string, isExpanded: boolean): void;
+  onConnectionSelect?(id: string): void;
+  onDatabaseExpand(id: string, isExpanded: boolean): void;
+  onNamespaceAction(namespace: string, action: Actions): void;
+  activeNamespace?: string;
+  isReadOnly?: boolean;
+}> = ({
+  connections,
+  databasesLegacy,
+  expanded = {},
+  expandedLegacy = {},
+  activeNamespace = '',
+  // onConnectionExpand and onConnectionSelect only have a default to support legacy usage
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  onConnectionExpand = () => {},
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  onConnectionSelect = () => {},
+  onDatabaseExpand,
+  onNamespaceAction,
+  isReadOnly = false,
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-expect-error ignoring test props so they are not part of the interface
+  __TEST_REACT_AUTOSIZER_DEFAULT_WIDTH = null,
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-expect-error ignoring test props so they are not part of the interface
+  __TEST_REACT_AUTOSIZER_DEFAULT_HEIGHT = null,
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-expect-error ignoring test props so they are not part of the interface
+  __TEST_REACT_WINDOW_OVERSCAN = null,
+}) => {
+  const listRef = useRef<List | null>(null);
+  const id = useId();
+
+  const isLegacy = !!databasesLegacy;
+
+  useEffect(() => {
+    if (activeNamespace) {
+      onDatabaseExpand(activeNamespace, true);
+    }
+  }, [activeNamespace, onDatabaseExpand]);
+  // TODO(COMPASS-7775): the we'll need something similar to expand the active connection
+
+  const items: TreeItem[] = useMemo(() => {
+    if (connections) {
+      return connections
+        .map((connection, connectionIndex) =>
+          connectionToItems({
+            connection,
+            connectionIndex,
+            connectionsLength: connections.length,
+            expanded,
+          })
+        )
+        .flat();
+    } else {
+      return databasesLegacy! // either connections or databasesLegacy should be set
+        .map((database, databaseIndex) =>
+          databaseToItems({
+            database,
+            databaseIndex,
+            databasesLength: databasesLegacy?.length || 0,
+            expanded: expandedLegacy,
+            level: 1,
+          })
+        )
+        .flat();
```

Comment:
Not a huge difference, but just fyi, there's also `.flatMap()` which does `.map().flat()` as a single call :)

Reply:
neat!